### ISSUE TITLE
Proposal to get rid of wkhtmltopdf and replace with a clean, small library

### ISSUE THREAD

**Issue by jpaulin** [odoo/odoo#21255]
Impacted versions: 
future versions of Odoo

Current behavior:
installation of Odoo is unnecessarily complicated by the requirement to install 'wkhtmltox' or 'wkhtmltopdf', which are slightly doing the "tail wagging the dog".

Expected behavior:
installation of Odoo could be properly done with a script that is completely unattended **and** works reliably across the Linux-sphere.

Upvote if you have had particularly gnawing (time consuming) problems with getting
wkhtmltopdf up and running.

I'm just interested in 2 things:
a) is wkhtmltopdf a necessary part of Odoo
b) are there clear benefits in terms of making Odoo more user-friendly (rather, "admin-friendly" is the correct word) by swapping 'wkhtmltopdf' with some other library

**Comment by Elkasitu** [comment]
Hi, replacing wkhtmltopdf is not a priority within Odoo right now, however we're open to suggestions or even better, a PR.

We will probably phase out wkhtmltopdf at some point, just not right now

**Comment by gdgellatly** [comment]
@Elkasitu My suggestion would be to evaluate and follow these 2 projects.  Both would allow the reports to remain largely untouched as at their core they are htmltopdf converters.

Simpler, with better layout, could use some help and support from an interested editor - https://github.com/xhtml2pdf/xhtml2pdf
- It is pure python with early P3 support.
- Supports Django report generation.
- More print quality focused (e.g. placement on page, pagination etc)
I toyed with replacing wkhtmltopdf in v7 with this as pagination and layout was a nightmare compared to RML.  While it wasn't quite drop in back then, that was before qweb and I suspect it is easier now.

Headless Chrome printing - something like
https://github.com/OCA/reporting-engine/pull/159
- obviously heavier, much more untested, but should evolve as chrome evolves.

And finally
As a simple suggestion for most problems - update the wkhtmltopdf documentation and recommendations.  12.4 has been relatively successful for most people.  But it still has issues with dpi settings and zoom, and is a little fiddly to get exactly right if you are unlucky enough to hit these issues.  

However it also shows the problem with wkhtmltopdf, which is no more love.  These dpi bugs have been reported and going on for years.

**Comment by ceffylau** [comment]
I also agree that this should be phased out. It seams like an unnecessary dependency and a fully integrated solution would be much better.

**Comment by ashkulz** [comment]
I'm the maintainer for wkhtmltopdf. I'd appreciate feedback if the distro-specific packages [for 0.12.5-dev](https://builds.wkhtmltopdf.org/0.12.5-dev/) fix all outstanding issues ... if not, I'm open to reverting to the 0.12.1 behavior for DPI as there seems breakage either way.

**Comment by Yenthe666** [comment]
@mart-e @sle-odoo and @odony can I ask one of you guys to look into the `wkhtmltopdf` issues for `0.12.5`? I was testing yesterday with Ubuntu `18.04` and it looks like `wkhtmltopdf` version `0.12.1` will not work there anymore since one of the dependencies is not available on Ubuntu 18.04 (See https://github.com/wkhtmltopdf/wkhtmltopdf/issues/3898 and https://packages.ubuntu.com/search?keywords=libpng12-0)
The package `libpng12-0` has no support for Ubuntu 18.04. The author of wkhtmltopdf (@ashkulz) is also mentioning that we are three versions behind so perhaps it is time to take a leap forward, atleast for Odoo `12.0`?

**Comment by pedrobaeza** [comment]
I think the plan is to switch to Chrome headless PDF rendering.

**Comment by mart-e** [comment]
@pedrobaeza are you sure you are not talking about chrome headless to replace phantomjs?
We have no plan to switch to another engine right now (but are not closed to the discussion).

Thanks @ashkulz for the feedback, I appreciate the follow-up. I will see with @d-fence if we can give a proper test case to explain exactly what works and what does not.
Sticking with an old version is not beneficial and is probably responsible to most complain about the difficulty to deploy it. On the docker image, we [currently ship](https://github.com/odoo/docker/blob/b3d6d018f740ad2a816b5209300b875e2760085e/11.0/Dockerfile#L19) the 0.12.4 but we don't get many feedback from there.

**Comment by ashkulz** [comment]
Till that happens, I've created a [branch](https://github.com/wkhtmltopdf/wkhtmltopdf/commits/0.12.1.x) with compilation fixes on top of 0.12.1 and produced binaries for  [`0.12.1.3`](https://builds.wkhtmltopdf.org/0.12.1.3/). I used the scripts in the [new packaging repository](https://github.com/wkhtmltopdf/packaging) to produce it, so if you have an OS/distro combination you'd like to see please open an issue there.

_Please note that the URLs are likely to change in future, so would appreciate not using them in CI scripts_.

**Comment by pedrobaeza** [comment]
@mart-e yes I should mix both things when talking about Chrome headless. FYI, we are shipping 0.12.4 version in our Docker image since the beginning: https://github.com/Tecnativa/docker-odoo-base. The problem is that we have to redesigned all the report styles for matching the new proportions. We don't have very heavy customization layers in the reports, so this wasn't be hard for us, but maybe for others can be a problem. And now if we switch back to the previous behavior, we will be forced again to "revert" changes (they are not exactly a revert, because some of the custom designs have been done directly on 0.12.4 version). @ashkulz is there any possibility of having a flag for preserving 0.12.4 behavior in newest 0.12.5.1 build?

**Comment by ashkulz** [comment]
@pedrobaeza: I'm quite confused, because Odoo recommends 0.12.1 and that's why I created 0.12.1.3 as a stopgap measure. I haven't reverted the 0.12.4 behavior yet, so you can test the 0.12.5-dev builds and see how that works out. I'm still on the fence for that, trying to figure what will cause the least amount of damage as both cannot be kept together (it's a patch in the custom Qt).

**Comment by pedrobaeza** [comment]
Yeah, as stated, officially Odoo is still sticked to 0.12.1. My company, that is an Odoo integrator, diverges from this for having a proper build over the Docker images we use, with version 0.12.4, with that side effect. Obviously, if there's no chance of having both possibilities, and you decide to patch again Qt with the same behavior as in 0.12.1, we will respect that decision and adapt our flows/reports to that.

Any way, the call is for Odoo itself. @mart-e is a good representative of them, but I suppose they have to discuss internally (although for sure the call to have the same behavior with updated binaries is very attractive for them).

**Comment by d-fence** [comment]
I did a print of the same sale order (demo data) with Odoo 11.0 and various versions of wkhtmltopdf
The first one should be the reference because it was made on the runbot with wk 0.12.1 (qt patched):
[Order_SO007_runbot_wkhtml_0.12.1_qt_patched.pdf](https://github.com/odoo/odoo/files/1967056/Order_SO007_runbot_wkhtml_0.12.1_qt_patched.pdf)

Here are the others:
[Order_SO007_docker_wkhtml_0.12.4.pdf](https://github.com/odoo/odoo/files/1967055/Order_SO007_docker_wkhtml_0.12.4.pdf)
[Order_SO007_wkhtmltox_0.12.1.3.pdf](https://github.com/odoo/odoo/files/1967057/Order_SO007_wkhtmltox_0.12.1.3.pdf)
[Order_SO007_wkhtmltox_0.12.5.pdf](https://github.com/odoo/odoo/files/1967058/Order_SO007_wkhtmltox_0.12.5.pdf)

It seems that 0.12.1.3 is the one that resemble the most to the reference.
The others are a bit too small.

@ashkulz it would be great to have a 0.12.1.3 official release that we could use in scripts (like the docker image) and have the link in our documentation

**Comment by ashkulz** [comment]
As 0.12.1.3 is 0.12.1 with a few compilation fixes, it is obviously going to resemble 0.12.1 the most (in fact, it should be pretty much exact). However, I don't think that remaining on an older version is a good strategy -- so that's why I asked what are the outstanding issues with 0.12.5-dev -- the DPI changes were aiming on getting the exact same rendering across platforms, which is not the case with 0.12.1 (it would vary across windows and linux).

**Comment by Yajo** [comment]
AFAICS from the attached reports, the difference between 0.12.1 and 0.12.4 reports is quite obvious, but from 0.12.4 to the upcoming 0.12.5 both seem identical.

So, do you think it should be safe to upgrade for us, that we already use 0.12.4? We're always in Linux.

Getting https://github.com/wkhtmltopdf/wkhtmltopdf/issues/2711 fixed is quite important!

**Comment by ashkulz** [comment]
There aren't any major changes in `0.12.5`, so I think it should be safe to upgrade.

I strongly feel that if the only pain is changing the report formats, a `--zoom` workaround can always be used -- but the latest version should be used, rather than trying to backport fixes to `0.12.1` forever.

**Comment by ashkulz** [comment]
wkhtmltopdf 0.12.5 has been released.

**Comment by sixtyfive** [comment]
Pleeeeeeeeeeeeeeeeease let this happen, dear gods!

**Comment by jpaulin** [comment]
Hi all!

How do you feel about this feature itself; keeping the issue open useful? (I am personally currently
not very active in Odoo issues in general). Thanks for any feedback!

**Comment by jeevan-smartltd** [comment]
> Hi all!
> 
> How do you feel about this feature itself; keeping the issue open useful? (I am personally currently not very active in Odoo issues in general). Thanks for any feedback!

If you could give a try to this [plutoprint ]( https://github.com/jeevanism/pludooprint)implementation and give the feedback, that would be great

**Comment by mart-e** [comment]
Hi there,
Thanks for sharing pludooprint, that’s nice to see community projects like this.

FYI, we have an ongoing effort to develop our own engine but it is still under heavy development
https://github.com/odoo/paper-muncher/

**Comment by kmagusiak** [comment]
Duplicate of #86501
